Library/Homebrew/cmd/list.rb

```diff
@@ -153,21 +169,41 @@ def filtered_list(args:)
       pinned_versions = {}
       names.sort.each do |d|
         keg_pin = (HOMEBREW_PINNED_KEGS/d.basename.to_s)
-        pinned_versions[d] = keg_pin.readlink.basename.to_s if keg_pin.exist? || keg_pin.symlink?
+        next if !keg_pin.exist? && !keg_pin.symlink?
+
+        keg = keg_pin.realpath
+        next unless select_keg?(keg, args)
+
+        pinned_versions[d] = keg.basename.to_s
       end
       pinned_versions.each do |d, version|
         puts d.basename.to_s.concat(args.versions? ? " #{version}" : "")
       end
     else # --versions without --pinned
       names.sort.each do |d|
-        versions = d.subdirs.map { |pn| pn.basename.to_s }
+        versions = d.subdirs.map { |pn| pn.basename.to_s if select_keg?(pn, args) }.compact
+        next if versions.empty?
         next if args.multiple? && versions.length < 2
 
         puts "#{d.basename} #{versions * " "}"
       end
     end
   end
 
+  def select_formula?(formula, args)
+    select_tab?(args) { Tab.for_formula(formula) }
+  end
+
+  def select_keg?(keg, args)
+    select_tab?(args) { Tab.for_keg(keg) }
+  end
+
+  def select_tab?(args)
+    return true if !args.installed_on_request? && !args.installed_as_dependency?
+
+    [args.installed_on_request?, !args.installed_as_dependency?].include?(yield.installed_on_request.present?)
```

Comment:
Actually, you should be able to solve the problem I mentioned above by doing this which I think is easier to understand, anyway.

```suggestion
    (args.installed_on_request? && tab.installed_on_request.present?) ||
      (args.installed_as_dependency? && tab.installed_as_dependency.present?)
```